### DISCUSSION TITLE
New C# keyword - "every"

### DISCUSSION THREAD

**Discussion by MillerMark** [dotnet/csharplang#5918] [Language Ideas]
I’m looking at code like this:

            var myKeepAliveTask = Task.Run(async () =>
            {
                while (true)
                {
                    await client.CustomerApi.Account.KeepStreamAliveAsync(key.Data);
                    await Task.Delay(TimeSpan.FromMinutes(30));
                }
            });

And I’m thinking the intent of this code would be easier to read (and code like it would be simpler to write) if it looked like this:

            every (TimeSpan.FromMinutes(30))
                await client.CustomerApi.Account.KeepStreamAliveAsync(key.Data);

I'm not a language expert so my apologies if this is a horribly bad suggestion.

**Comment by iam3yal** [comment]
You can do it today, introduce a method and name it `every` in your codebase but really I why do you have an infinite loop around an async code, doesn't it defeat the purpose? sounds like this should be `Task.Factory.StartNew(() => { }, TaskCreationOptions.LongRunning)`.

You need to ask yourself in the grand scheme of things how this fits in the language? it's not that useful and too niche to be considered as feature not to mention that you haven't provided the exact semantics but if the point is to make it readable then I'd argue that the original code above while not shorter is more readable and more transparent and for a good reason it shouldn't be a keyword.

**Reply by Xyncgas** [reply]
Comes in the guy who runs a factory

**Reply by Xyncgas** [reply]
it's ok I get it
```C#
async Task<bool> every (int ms) => await Task.Delay(ms);
async Task Main()
{
    while(await every(1000))
        await EatTwoScoopsOfIceCream(); //Do something else while you wait for joe to finish his ice creams
}
```

**Reply by iam3yal** [reply]
@Xyncgas I'm sure you know that by now but if you don't have anything constructive to say it's better to keep your thoughts to yourself and save it from the rest of the world and no unfortunately you didn't get what I was saying. Thank you, have a nice day.

I wasn't trying to be sarcastic but really it shouldn't be `asynchronous` so just to clarify this is what I meant with my original comment:

```csharp
static async Task every(TimeSpan delay, Func<Task> func)
{
    while (true)
    {
        await func();
        await Task.Delay(delay);
    }
}
```
or (depends on what the OP really meant.)
```csharp
static Task every(TimeSpan delay, Func<Task> func)
{
    return Task.Run(async () =>
    {
        while (true)
        {
            await func();
            await Task.Delay(delay);
        }
    });
}
 ```

**Comment by jnm2** [comment]
Also consider that as written, that infinite loop aborts and never comes back if KeepStreamAliveAsync throws an exception. This would not be the case when directly using [`System.Threading.Timer`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.timer) (which underlies `Task.Delay`).
